criu/kerndat.c

```diff
@@ -226,6 +235,46 @@ static int kerndat_get_shmemdev(void)
 	return -1;
 }
 
+static int kerndat_get_hugetlb_dev(void)
+{
+	void *map;
+	int i, flag;
+	unsigned long long size;
+	dev_t dev;
+
+	for (i = 0; i < HUGETLB_MAX; i++) {
+		/* This hugetlb size is not supported or the dev's number has been collected */
+		if (kdat.hugetlb_dev[i])
+			continue;
+
+		size = hugetlb_info[i].size;
+		flag = hugetlb_info[i].flag;
+		map = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | flag, 0, 0);
+		if (map == MAP_FAILED && errno == EINVAL) {
+			kdat.hugetlb_dev[i] = (dev_t)-1;
+			continue;
+		}
+
+		if (map == MAP_FAILED && errno == ENOMEM) {
+			pr_info("Hugetlb size %llu Mb is supported but cannot get dev's number\n", size >> 20);
```

Comment:
Here `kdat.hugetlb_dev[i]`will be 0. Is this what you expect?

Reply:
Yes, it is what I want. `kdat.hugetlb_dev[i]` will be -1 if we detect that hugetlb size is not supported, 0 if it is supported but we cannot mmap (this is based on my assumption that the major, minor number cannot be both 0, I didn't look carefully at this https://github.com/torvalds/linux/blob/master/fs/super.c#L1032). Each time `kerndat_get_hugetlb_dev()` gets called, it only look at `kdat.hugetlb_dev[i] == 0` to try to get the device number of the hugetlb size. I will add comment on this.